### ISSUE TITLE
[enhancement] add C style casts

### ISSUE THREAD

**Issue by A60AB5450353F40E** [CashScript/cashscript#155]
Currently some casts will actually call conversion like int() resulting in compiled bytecode using OP_BIN2NUM.

In some cases, when the user knows that the hex will be a valid script number this will be terribly wasteful.

Consider this:

```
int a=1;
bytes1 b=0x01;
require((bytes) a == b);
```
it would compile to `OP_1 OP_1 OP_EQUALVERIFY`

while

```
int a=1;
bytes1 b=0x01;
require(a == (int) b);
```

would compile to `OP_1 OP_1 OP_NUMEQUALVERIFY`.

What if b is not a valid int encoding? Then still compile the same but compiled Script would fail due to OP_NUMEQUALVERIFY erroring.

**Comment by rkalis** [comment]
I agree that there should probably be a way to "just" typecast without adding any opcodes, but I'm not sure if I like having two separate "casting" syntaxes. The only cast where this applies is casting to `int` (when casting `int` to e.g. `bytes4` you can already do `bytes1(bytes(20))` to forego the `OP_NUM2BIN` (if you wanted to). So we may want to allow something similar for `int` casting. This is not a big priority right now, but I think it also shouldn't be a lot of work, so I'll try to think of some syntax that makes sense.

**Comment by mr-zwets** [comment]
how does the `bool()` cast coerce values to either 0 or 1?

or is the current `bool()` cast an unsafe semantic cast?

**Comment by rkalis** [comment]
Yeah `bool()` is currently just a semantic cast. The only non-semantic cast we have now is between int and bounded bytes (num2bin/bin2num). We've chatted about this before, and I do think it makes sense to have two separate syntaxes for semantic casting and actual data conversions.

**Comment by rkalis** [comment]
Currently these are the existing casting behaviours:
```
bytes(int) => semantic
bytes6(int) => padding to 6 bytes (or fails if too big)
padInt(int, 6);
bool(int) => semantic

int(bool) => semantic
bytes(bool)) => not allowed

bytes(string) => semantic
bytes6(string) => not allowed

sig(bytes) => semantic
pubkey(bytes) => semantic
datasig(bytes) => semantic
int(bytes) => minimally encode as number
bytes6(bytes) => semantic

bytes(pubkey | sig | datasig) => semantic
bytes6(pubkey | sig | datasig) => not allowed
```

We want there to be a separate syntax for semantic `int => bytes` casting and "padding" `int => bytes` casting. Also for semantic `bytes => int` casting and "number encoding" `int => bytes` casting. For `int => bool` we need there to be *a* functionality where the value *does* change to `OP_1` or `OP_0`, rather than only being semantic.

The biggest thing is the fact that currently `bool(5) == true` evaluates to `false`. I don't think there's any other programming language where that would be the case. So, we want the "normal" casting behaviour of `bool(int)` to coerce the value to `OP_1` or `OP_0`.

Then, some people, like @A60AB5450353F40E who opened this issue, might want to skip this extra operation because they already know that the value they are casting to is already `1` or `0`, which would be an unsafe bool cast (equivalent to the current behaviour).

We should do a similar thing for `int()` casting, which is the issue that @A60AB5450353F40E raised in the first place. We want to keep the current "minimal encoding" behaviour as the default, but also add an "unsafe int cast" that just assumes you know what you're doing and doesn't add an operation.

For `bytes => bytesX` casting, there is really no "safe" way to do so, so we should rename this operation to `unsafe_bytesX()`.

For the "padding" behaviour, we should rename that to be explicit about the fact that padding is going on. Something like `padInt(int, 6)`. And then completely remove the `bytes6(int)` or `bytes(int, 6)` syntax. We should also consider whether there is a way to add padding to a `bytes` (doesn't look like there is an easy way for this right now, but we can do it the hard way).

**Comment by rkalis** [comment]
This is the list of TODOs:
- [x] Fix default `bool()` casting
- [x] Add `unsafe_bool()` casting
- [x] Add `unsafe_int()` casting
- [x] Rename `bytesX()` casting to `unsafe_bytesX()` casting
- [x] Disallow `int => bytesX` casting
- [x] Rename `bytes(int, 6)` to `toPaddedBytes(int, 6)`
- [ ] ~Add `toPaddedBytes(bytes, 6)` functionality~

**Comment by rkalis** [comment]
Today, we managed to tick off almost everything on this list, except for the (optional) `toPaddedBytes(bytes, 6)` functionality. On Thursday we will look further into this.

We also still need to add a few small tests for `unsafe_bool` and `unsafe_int`, and update the casting docs once everything is updated.

**Comment by rkalis** [comment]
We discussed `toPaddedBytes(bytes, 6)` today but it doesn't seem like there is a real use case for it. If some `bytes` data would need to be encoded for NFT commitment (or similar), then there are probably better encoding methods than padding with 00s. In the future we probably want to add first-class language support for encoding/decoding data from NFT commitments.

For now, we're not adding a padidng functionality for `bytes` types.
